kernel/kernel-default/src/main/java/io/delta/kernel/client/DefaultJsonHandler.java

```diff
@@ -0,0 +1,187 @@
+                }
+
+                // There is no file in reading or the current file being read has no more data
+                // initialize the next file reader or return false if there are no more files to
+                // read.
+                try {
+                    if (currentFileReader == null ||
+                        (nextLine = currentFileReader.readLine()) == null) {
+
+                        tryOpenNextFile();
+                        if (currentFileReader != null) {
+                            nextLine = currentFileReader.readLine();
+                        }
+                    }
+                }
+                catch (IOException ex) {
+                    throw new RuntimeException(ex);
+                }
+
+                return nextLine != null;
+            }
+
+            @Override
+            public FileDataReadResult next()
+            {
+                if (nextLine == null) {
+                    throw new NoSuchElementException();
+                }
+
+                List<Row> rows = new ArrayList<>();
+                int currentBatchSize = 0;
+                do {
+                    // hasNext already reads the next one and keeps it in member variable `nextLine`
+                    rows.add(parseJson(nextLine, physicalSchema));
+                    nextLine = null;
+                    currentBatchSize++;
+                }
+                while (currentBatchSize < maxBatchSize && hasNext());
+
+                ColumnarBatch batch = new DefaultRowBasedColumnarBatch(physicalSchema, rows);
+                Row scanFileRow = currentFile.getScanFileRow();
+                return new FileDataReadResult()
+                {
+                    @Override
+                    public ColumnarBatch getData()
+                    {
+                        return batch;
+                    }
+
+                    @Override
+                    public Row getScanFileRow()
+                    {
+                        return scanFileRow;
+                    }
+                };
+            }
+
+            private void tryOpenNextFile()
+                throws IOException
+            {
+                Utils.closeCloseables(currentFileReader); // close the current opened file
+                currentFileReader = null;
+
+                if (fileIter.hasNext()) {
+                    currentFile = fileIter.next();
+                    FileStatus fileStatus =
+                        Utils.getFileStatus(currentFile.getScanFileRow());
+                    Path filePath = new Path(fileStatus.getPath());
+                    FileSystem fs = filePath.getFileSystem(hadoopConf);
+                    FSDataInputStream stream = null;
+                    try {
+                        stream = fs.open(filePath);
+                        currentFileReader = new BufferedReader(
+                            new InputStreamReader(stream, StandardCharsets.UTF_8));
+                    } catch (Exception e){
```

Comment:
```suggestion
                    } finally {
```
?

Reply:
not needed with silent close?

Reply:
Yep and this would be wrong anyways since we don't want to close the stream..